### PR TITLE
Graph quantization

### PR DESCRIPTION
**Pull request by Giuseppe5**
Address #464

### CODE REVIEW

File: src/brevitas/graph/target/flexml.py [hunk 1]
```diff
@@ -78,9 +78,14 @@
         'act_quant': Uint8ActPerTensorFixedPointMaxInit,
         'max_val': 6.,
         'return_quant_tensor': True}),
-    nn.LeakyReLU: (qnn.flexml.FlexMLQuantLeakyReLU, {})
+    nn.LeakyReLU: (qnn.flexml.FlexMLQuantLeakyReLU, {}),
+    nn.Hardtanh: (qnn.QuantHardTanh, {
+        'act_quant': Int8ActPerTensorFixedPoint,
```
Comment (Giuseppe5) [inline src/brevitas/graph/target/flexml.py:83]: Int8ActperTensorFixedPointFromMinMax

Comment (Giuseppe5) [reply src/brevitas/graph/target/flexml.py:83]: Use lambda to retrieve attributes from original NN module to QuantModule

File: src/brevitas/graph/target/flexml.py [hunk 2]
```diff
@@ -159,7 +164,7 @@ def are_inputs_unsigned(model, node, is_unsigned_list):
                 is_unsigned_list.append(True)
             elif isinstance(inp_module, tuple(SIGN_PRESERVING_MODULES)):
                 are_inputs_unsigned(model, inp_node, is_unsigned_list)
-            elif isinstance(inp_module, (qnn.QuantReLU, qnn.QuantIdentity, qnn.QuantHardTanh)):
+            elif isinstance(inp_module, QUANT_ACT_MODULES):
```
Comment (Giuseppe5) [inline src/brevitas/graph/target/flexml.py:167]: check interface (if hasattr(inp_module, 'is_quant_act_signed')

File: src/brevitas/graph/standardize.py [hunk 3]
```diff
@@ -81,7 +81,7 @@ def rewrite_node(self, node: Node, graph_model: GraphModule):
         with graph_model.graph.inserting_after(batch_size_node):
             squeeze_node = graph_model.graph.call_method(
                 'reshape', args=(node, (batch_size_node, -1)))
-        replace_all_uses_except(node, squeeze_node, [squeeze_node, batch_size_node])
+        replace_uses_except(node, squeeze_node, [squeeze_node, batch_size_node])
```
Comment (volcacius) [inline src/brevitas/graph/standardize.py:84]: Can we use the implementation as by leveraging except?

**Review by volcacius — COMMENTED** [review]
Move the graph quantization functions to graph/quantize.py and rename them to remove references to flexml.
Add quantize() function in graph/quantize.py that is parametrized with the dicts, with per-tensor float scales as defaults, calling:
graph_model = inp_placeholder_handler(graph_model, input_quantizer=input_quantizer)
    graph_model = act_handler(graph_model, quant_act_map)
    graph_model = add_output_quant_handler(graph_model, quant_identity_map, quant_act_map, unsigned_act_tuple)
    graph_model = wbiol_handler(graph_model, quant_wbiol_map, quant_identity_map)
  graph_model = DisableLastReturnQuantTensor().apply(graph_model)

Then under target/flexml.py keep quantize_flexml(), but calling quantize() inside.

File: src/brevitas/graph/quantize.py [hunk 4]
```diff
@@ -0,0 +1,524 @@
...
+    nn.AvgPool3d,
+    nn.AdaptiveAvgPool1d,
+    nn.AdaptiveAvgPool2d,
+    nn.AdaptiveAvgPool3d]
+
+
+QUANT_WBIOL_MAP = {
+    nn.Conv1d: (qnn.QuantConv1d, {
+        'weight_quant': Int8WeightPerTensorFloat,
+        'bias_quant': Int16Bias,
```
Comment (Giuseppe5) [inline src/brevitas/graph/quantize.py:53]: Int32Bias

File: src/brevitas/nn/target/flexml.py [hunk 5]
```diff
@@ -48,6 +48,77 @@ def act_quant(self):
     def is_quant_act_signed(self):
         return self.output_quant.is_quant_act_signed
 
+
+class FlexMLQuantSwish(nn.Module):
```
Comment (Giuseppe5) [inline src/brevitas/nn/target/flexml.py:52]: Change to QuantRelu-like implementation

File: src/brevitas/graph/quantize.py [hunk 6]
```diff
@@ -0,0 +1,524 @@
...
+        'output_quant': qnn.QuantIdentity(Int8ActPerTensorFloat, return_quant_tensor=True)}),
+    nn.Hardtanh: (qnn.QuantHardTanh, {
+        'act_quant': Int8ActPerTensorFloatMinMaxInit,
+        'max_val': lambda module: module.max_val,
+        'min_val': lambda module: module.min_val
+    }),
+    nn.Sigmoid: (qnn.QuantSigmoid, {
+        'act_quant': Uint8ActPerTensorFloat,
+        'return_quant_tensor': True,}),
+    nn.SiLU: (qnn.flexml.FlexMLQuantSwish, {
```
Comment (Giuseppe5) [inline src/brevitas/graph/quantize.py:102]: flexml-only act to be included only in FlexML config

File: src/brevitas/graph/target/flexml.py [hunk 7]
```diff
@@ -1,432 +1,171 @@
...
+            'alpha_quant':
+                qnn.QuantIdentity(Uint8ActPerTensorFixedPoint, bit_width=16),
+            'input_quant':
+                qnn.QuantIdentity(
+                    Int8ActPerTensorFixedPoint, bit_width=16, scaling_stats_momentum=None),
+            'output_quant':
+                qnn.QuantIdentity(Int8ActPerTensorFixedPoint, return_quant_tensor=True)}),
+    nn.Hardtanh: (
+        qnn.QuantHardTanh,
+        {
```
Comment (volcacius) [inline src/brevitas/graph/target/flexml.py:89]: why is this on a new line but not the others?

File: src/brevitas/graph/target/flexml.py [hunk 8]
```diff
@@ -1,432 +1,171 @@
...
+    'unsigned': (
+        qnn.QuantIdentity, {
+            'act_quant': Uint8ActPerTensorFixedPoint, 'return_quant_tensor': True}),}
 
 
 def preprocess_flexml(
         model,
         *model_args,
+        trace_model=True,
         equalization_iters=0,
```
Comment (volcacius) [inline src/brevitas/graph/target/flexml.py:123]: equalize_iters

File: src/brevitas/graph/target/flexml.py [hunk 9]
```diff
@@ -1,432 +1,171 @@
...
         merge_bias=True,
+        merge_bn=True,
         bias_shrinkage='vaiq',
```
Comment (volcacius) [inline src/brevitas/graph/target/flexml.py:126]: equalize_

File: src/brevitas/graph/quantize.py [hunk 10]
```diff
@@ -0,0 +1,598 @@
...
+                quant_module_class, quant_module_kwargs = float_to_quant_wbiol_map[type(module)]
+                rewriter = ModuleToModuleByInstance(
+                    module, quant_module_class, **quant_module_kwargs)
+                rewriters.append(rewriter)
+    for rewriter in rewriters:
+        model = rewriter.apply(model)
+    return model
+
+
+def avgpool_handler(
```
Comment (volcacius) [inline src/brevitas/graph/quantize.py:525]: add a flag requantize_output, that is then exposed at the top of quantize as requantize_avgpool_output

**Review by volcacius — CHANGES_REQUESTED** [review]

File: src/brevitas/graph/quantize_impl.py [hunk 11]
```diff
@@ -0,0 +1,423 @@
...
+            if (node.op == 'call_function' and node.target in ADD_FNS + [CAT] or
+                    node.op == 'call_method' and node.target in ADD_METHODS):
+                rewriters = []
+                # If the op is CAT, check that inputs have same sign, and in recursive_input_handler
+                # force that the sign is aligned
+                same_sign = node.target is CAT
+
+                # If input to the CAT or ADD node are quantized and aligned correctly, continue to
+                # the next node
+                if are_inputs_quantized_and_aligned(model,
```
Comment (volcacius) [inline src/brevitas/graph/quantize_impl.py:303]: is the formatting ok here?

File: src/brevitas/graph/target/flexml.py [hunk 12]
```diff
@@ -1,432 +1,177 @@
...
+
+FLEXML_COMPUTE_LAYER_MAP = {
+    nn.AvgPool2d: (qnn.flexml.FlexMLQuantAvgPool2d, {
+        'return_quant_tensor': True}),
+    nn.MultiheadAttention: (
+        qnn.QuantMultiheadAttention,
+        {
+            'in_proj_weight_quant': Int8WeightPerTensorFixedPoint,
+            'in_proj_bias_quant': Int16Bias,
+            'attn_output_weights_quant': Int8ActPerTensorFixedPoint,
```
Comment (volcacius) [inline src/brevitas/graph/target/flexml.py:35]: unsigned

**Review by volcacius — CHANGES_REQUESTED** [review]

File: src/brevitas/graph/quantize_impl.py (hunk 11 above)
Comment (Giuseppe5) [reply src/brevitas/graph/quantize_impl.py:303]: Maybe the function call inside an _if_ is different than a normal function call

**Review by volcacius — APPROVED** [review]
LGTM

**Pull request merged**